lib/Modules/HyperVCIHooks/HyperVCIHooks.psm1

```diff
@@ -0,0 +1,1538 @@
+    }
+  
+    $networkType = Get-JujuCharmConfig -Scope 'network-type'
+    Initialize-GitRepositories $networkType $BranchName $BuildFor
+
+    $standardProjects = 'openstack/nova', 'openstack/neutron', 'openstack/os-win'
+    if ($buildFor -in $standardProjects) {
+        $defaultProjects = $standardProjects -notmatch $buildFor
+        foreach ($project in $defaultProjects) {
+            Install-Projects $project
+        }
+        Install-NetworkType
+        Install-Projects $buildFor
+    }
+    else {
+        foreach ($project in $standardProjects) {
+            Install-Projects $project
+        }
+        if (($buildFor -eq "none") -or ($buildFor -eq "openstack/networking-hyperv")) {
+            Install-NetworkType
+        }
+        else {
+            Install-NetworkType
+            Install-Projects $buildFor
+        }
+    }
+
+    Write-JujuLog "Environment initialization done."
+}
+
+function Set-ServiceAcountCredentials {
+    Param(
+        [string]$ServiceName,
+        [string]$ServiceUser,
+        [string]$ServicePassword
+    )
+
+    $filter = 'Name=' + "'" + $ServiceName + "'" + ''
+    $service = Get-WMIObject -Namespace "root\cimv2" -Class Win32_Service -Filter $filter
+    $service.StopService()
+    while ($service.Started) {
+        Start-Sleep -Seconds 2
+        $service = Get-WMIObject -Namespace "root\cimv2" -Class Win32_Service -Filter $filter
+    }
+
+    Grant-Privilege $ServiceUser "SeServiceLogonRight"
+    Set-ServiceLogon -Services @($ServiceName) -UserName $ServiceUser -Password $ServicePassword
+}
+
+
+function New-OpenStackService {
+    Param(
+        [string]$ServiceName,
+        [string]$ServiceDescription,
+        [string]$ServiceExecutable,
+        [string]$ServiceConfig,
+        [string]$ServiceUser,
+        [string]$ServicePassword
+    )
+
+    $filter='Name=' + "'" + $ServiceName + "'"
+
+    $service = Get-WmiObject -Namespace "root\cimv2" -Class Win32_Service -Filter $filter
+    if($service) {
+        Write-JujuLog "Service $ServiceName is already created."
+        return $true
+    }
+
+    $serviceFileName = "OpenStackService.exe"
+    if(!(Test-Path "$SERVICE_DIR\$serviceFileName")) {
+        Copy-Item "$FILES_DIR\$serviceFileName" "$SERVICE_DIR\$serviceFileName"
+    }
+
+    New-Service -Name "$ServiceName" -DisplayName "$ServiceName" -Description "$ServiceDescription" -StartupType "Manual" `
+                -BinaryPathName "$SERVICE_DIR\$serviceFileName $ServiceName $ServiceExecutable --config-file $ServiceConfig" `
+
+    if((Get-Service -Name $ServiceName).Status -eq "Running") {
+        Stop-Service $ServiceName
+    }
+
+    Set-ServiceAcountCredentials $ServiceName $ServiceUser $ServicePassword
+}
+
+
+function Watch-ServiceStatus {
+    Param(
+        [string]$ServiceName,
+        [int]$IntervalSeconds
+    )
+
+    $count = 0
+    while ($count -lt $IntervalSeconds) {
+        if ((Get-Service -Name $ServiceName).Status -ne "Running") {
```

Comment:
This will throw an error on first iteration of the while loop if the service is not running. Is this the intended outcome? If not, you should probably change it to something like:

```powershell
while ($count -lt $IntervalSeconds) {
    if ((Get-Service -Name $ServiceName).Status -ne "Running") {
        $count += 1
        Start-Sleep -Seconds 1
        continue
    }
    return
}
Throw "$ServiceName has errors. Please check the logs."
```